x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.support;
+
+import org.elasticsearch.action.admin.cluster.shards.ClusterSearchShardsAction;
+import org.elasticsearch.action.admin.indices.mapping.get.TransportGetFieldMappingsIndexAction;
+import org.elasticsearch.action.explain.TransportExplainAction;
+import org.elasticsearch.action.get.TransportGetAction;
+import org.elasticsearch.action.get.TransportShardMultiGetAction;
+import org.elasticsearch.action.search.TransportSearchShardsAction;
+import org.elasticsearch.action.support.TransportAction;
+import org.elasticsearch.action.termvectors.TermVectorsAction;
+import org.elasticsearch.action.termvectors.TransportShardMultiTermsVectorAction;
+import org.elasticsearch.common.inject.Binding;
+import org.elasticsearch.common.inject.TypeLiteral;
+import org.elasticsearch.index.seqno.RetentionLeaseActions;
+import org.elasticsearch.node.Node;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.xpack.core.LocalStateCompositeXPackPlugin;
+import org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.xpack.core.security.action.apikey.CrossClusterApiKeyRoleDescriptorBuilder.CCR_INDICES_PRIVILEGE_NAMES;
+import static org.elasticsearch.xpack.core.security.action.apikey.CrossClusterApiKeyRoleDescriptorBuilder.CCS_INDICES_PRIVILEGE_NAMES;
+
+public class CrossClusterShardTests extends ESSingleNodeTestCase {
+
+    Set<String> MANUALLY_CHECKED_SHARD_ACTIONS = Set.of(
+        // The request types for these actions are all subtypes of SingleShardRequest, and have been evaluated to make sure their
+        // `shards()` methods return the correct thing.
+        TransportShardMultiTermsVectorAction.TYPE.name(),
+        TransportExplainAction.TYPE.name(),
+        RetentionLeaseActions.ADD.name(),
+        RetentionLeaseActions.REMOVE.name(),
+        RetentionLeaseActions.RENEW.name(),
+        TermVectorsAction.NAME,
+        TransportGetAction.TYPE.name(),
+        TransportShardMultiGetAction.TYPE.name(),
+        TransportGetFieldMappingsIndexAction.TYPE.name(),
+
+        // These actions do not have any references to shard IDs in their requests.
+        ClusterSearchShardsAction.NAME,
+        TransportSearchShardsAction.NAME
+    );
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return Set.of(LocalStateCompositeXPackPlugin.class);
```

Comment:
This results in only checking the (core) xpack plugin and (core) server code's registered actions. Do you think it is worth pull this out to it's own qa project and have that qa project depend on a large % of the core/modules and x-pack/plugins. I understand this is best effort only but pulling this out to it's own project will get coverage similar to the default distribution and help to identify any new actions (unless the new actions are part of a new plugin/module)

Reply:
Blah, whoops, for some reason I thought switching this to `LocalStateCompositeXPackPlugin` addressed that but upon looking again you are correct and it does not. I'll look into this and look to get this pulled out somewhere we get more coverage.

(I really dislike this part of our test infrastructure and find it quite confusing, which plugins are included/accessible where)